### PR TITLE
Course deletion#7

### LINKED ISSUE #7: course deletion in user service
# Description
When a course is deleted, the user service needs to remove the course from all related entries in the database.

# Parent 
https://github.com/IT-REX-Platform/course_service/issues/29

# Acceptance criteria
- [ ] user service receives and handles message received from course service
- [ ] user service removes all entries related to the course

### PR DESCRIPTION
**Pull request by BlueWaves2**
## Description of changes
The user Service now removes all course memberships of courses that are deleted. Reception of deletion is received via dapr event.

Closes #7 
  
## How has this been tested?
Please describe the test strategy you followed.

- [X] automated unit test
- [ ] automated integration test
- [ ] automated acceptance test
- [ ] manual, exploratory test

    
## Checklist before requesting a review
- [X] My code follows the [coding guidelines](https://github.com/IT-REX-Platform/wiki/blob/main/dev-manuals/backend/coding-guidelines.md) of this project
- [X] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] My code fulfills all acceptance criteria
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] The test coverage (line and branch) is reasonably high, especially on Service classes and other classes with complex logic
- [X] I have made corresponding changes to the documentation
- [X] I have added explanation of architectural decision and rationales to [wiki/adr](https://github.com/IT-REX-Platform/wiki/tree/main/adr)
- [X] I have updated the changes in the ticket description

## Checklist for reviewer
- [ ] The code works and does not throw errors
- [ ] The code is easy to understand and there are no confusing parts
- [ ] The code follows the [coding guidelines](https://github.com/IT-REX-Platform/wiki/blob/main/dev-manuals/backend/coding-guidelines.md) of this project
- [ ] The code change accomplishes what it is supposed to do
- [ ] I cannot think of any use case in which the code does not behave as intended
- [ ] The added and existing tests reasonably cover the code change
- [ ] I cannot think of any test cases, input or edge cases that should be tested in addition
- [ ] Description of the change is included in the documentation

### CODE REVIEW

File: src/main/java/de/unistuttgart/iste/gits/user_service/service/MembershipService.java [hunk 1]
```diff
@@ -110,6 +112,39 @@ public CourseMembership deleteMembership(CourseMembershipInput inputDto) {
...
+    }
+
+    /**
+     * method to handle course deletion events. removes all course memberships associated with the course ID
+     * @param changeEvent course change event
+     */
+    public void removeCourse(CourseChangeEvent changeEvent){
+        // evaluate course Update message
+        if (changeEvent.getCourseId() == null || changeEvent.getOperation() == null){
+            throw new NullPointerException("incomplete message received: all fields of a message must be non-null");
```
Comment (PaulBredl) [inline src/main/java/de/unistuttgart/iste/gits/user_service/service/MembershipService.java:135]: Actually this is dangerous. When dapr calls a service with an event and this service returns 500 (or any other code other than 2XX), dapr will continue to call this service with the event (at-least-once semantics). So if for some reason, null is one of these fields, this will make dapr call this service with the malformed event until the end of time. (This is what currently happens in the reward service and it's a mess)

Here in the service it makes sense to throw an exception, but I would wrap the call in the subscription controller with try catch and catching that exception. Maybe change it to a "ValidationException", so not any null pointer exception is catched. 

Just my suggestion, maybe you have a better idea how to deal with it.

**Review by PaulBredl — CHANGES_REQUESTED** [review]
Only some dapr problem

**Review by PaulBredl — APPROVED** [review]
